arbos/programs/native.go

```diff
@@ -95,13 +114,57 @@ func activateProgram(
 		asmEstimate:   uint32(stylusData.asm_estimate),
 		footprint:     uint16(stylusData.footprint),
 	}
-	db.ActivateWasm(hash, asm, module)
-	return info, err
+	return info, asm, module, err
+}
+
+func getLocalAsm(statedb vm.StateDB, moduleHash common.Hash, address common.Address, pagelimit uint16, time uint64, debugMode bool, program Program) ([]byte, error) {
+	localAsm, err := statedb.TryGetActivatedAsm(moduleHash)
+	if err == nil && len(localAsm) > 0 {
+		return localAsm, nil
+	}
+
+	codeHash := statedb.GetCodeHash(address)
+
+	wasm, err := getWasm(statedb, address)
+	if err != nil {
+		log.Error("Failed to reactivate program: getWasm", "address", address, "expected moduleHash", moduleHash, "err", err)
+		return nil, fmt.Errorf("failed to reactivate program address: %v err: %w", address, err)
+	}
+
+	unlimitedGas := uint64(0xffffffffffff)
+	// we know program is activated, so it must be in correct version and not use too much memory
+	info, asm, module, err := activateProgramInternal(statedb, address, codeHash, wasm, pagelimit, program.version, debugMode, &unlimitedGas)
+	if err != nil {
+		log.Error("failed to reactivate program", "address", address, "expected moduleHash", moduleHash, "err", err)
+		return nil, fmt.Errorf("failed to reactivate program address: %v err: %w", address, err)
+	}
+
+	if info.moduleHash != moduleHash {
+		log.Error("failed to reactivate program", "address", address, "expected moduleHash", moduleHash, "got", info.moduleHash)
+		return nil, fmt.Errorf("failed to reactivate program. address: %v, expected ModuleHash: %v", address, moduleHash)
+	}
+
+	currentHoursSince := hoursSinceArbitrum(time)
+	if currentHoursSince > program.activatedAt {
+		// stylus program is active on-chain, and was activated in the past
```

Comment:
I'm pretty confused about what this check is trying to achieve, and when it wouldn't be true. First of all, is `time` the EVM time? Do note that can be overridden in an eth_call. If so, shouldn't the current EVM time always be greater than the program's activation time, or else it wouldn't be callable? How would it be activated in the future and saved in state as such?

Reply:
This is meant mostly to handle the case where a user does an eth_call which activates a wasm and calls it. We don't want to store a wasm activated in the same eth_call in our database because we'll probably never need it again.
However, if a user eth_calls to a wasm that's activated onchain - we do want to store the wasm to database, because it's activated onchain and gas was spent activating it.